File: scripts/process-parcel-data.py
#!/usr/bin/env python3
"""
process-parcel-data.py

Computes housing characteristics for 17 Seattle neighborhoods from King County
Assessor parcel data. Run once to generate housing-characteristics.json, which
is then hardcoded into neighborhoods.ts.

Data sources (downloaded from https://aqua.kingcounty.gov/extranet/assessor/):
  - Residential Building.zip  → EXTR_ResBldg.csv   (year built, garage sqft, ZIP)
  - Parcel.zip                → EXTR_Parcel.csv    (lot size, present use code)

Usage:
  python3 scripts/process-parcel-data.py

Looks for the ZIP files in scripts/.cache/ or /tmp/. Downloads them if missing.
"""

import csv
import json
import os
import sys
import urllib.request
import zipfile
from statistics import median as stat_median
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
CACHE_DIR = SCRIPT_DIR / ".cache"
OUTPUT = SCRIPT_DIR / "housing-characteristics.json"

RESBLDG_URL = "https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip"
PARCEL_URL  = "https://aqua.kingcounty.gov/extranet/assessor/Parcel.zip"
RESBLDG_ZIP = CACHE_DIR / "EXTR_ResBldg.zip"
PARCEL_ZIP  = CACHE_DIR / "EXTR_Parcel.zip"

# King County Parcel PresentUse codes (LUType 102 in LookUp table)
USE_SINGLE_FAMILY = {"2", "6", "9"}   # Single Family (various zones)
USE_TOWNHOUSE     = {"29"}            # Townhouse Plat
USE_CONDO         = {"20", "25"}      # Condo Residential, Condo Mixed Use

# ZIP → neighborhood. Where a ZIP spans multiple neighborhoods, all are listed.
# Source: Seattle neighborhood maps + ZIP code boundary overlays.
ZIP_MAP: dict[str, list[str]] = {
    "98109": ["Queen Anne"],
    "98119": ["Queen Anne"],
    "98107": ["Ballard"],
    "98177": ["Sunset Hill"],          # NW Seattle; Magnolia is 98199
    "98103": ["Wallingford", "Fremont", "Green Lake"],  # shared ZIP
    "98115": ["Roosevelt"],
    "98125": ["Maple Leaf / Ravenna"],
    "98117": ["Crown Hill"],           # Also NW; 98117 is primarily Crown Hill / north Ballard
    "98133": ["Phinney / Greenwood"],
    "98199": ["Magnolia"],
    "98112": ["Madison Valley"],
    "98102": ["Capitol Hill"],
    "98116": ["West Seattle (Admiral)"],
    "98118": ["Columbia City"],
    "98144": ["Mount Baker"],
    "98108": ["Beacon Hill"],
    "98106": ["Beacon Hill"],
}

ALL_NEIGHBORHOODS = sorted({n for names in ZIP_MAP.values() for n in names})


def download(url: str, dest: Path):
    CACHE_DIR.mkdir(parents=True, exist_ok=True)
    print(f"Downloading {url.split('/')[-1].replace('%20', ' ')} ...", flush=True)
    urllib.request.urlretrieve(url, dest)
    print(f"  → {dest} ({dest.stat().st_size / 1e6:.1f} MB)")


def ensure_cached():
    if not RESBLDG_ZIP.exists():
        # Check /tmp first (from previous download)
        tmp_resbldg = Path("/tmp/kc_resbldg.zip")
        if tmp_resbldg.exists():
            print(f"Using {tmp_resbldg}")
            CACHE_DIR.mkdir(parents=True, exist_ok=True)
            import shutil
            shutil.copy(tmp_resbldg, RESBLDG_ZIP)
        else:
            download(RESBLDG_URL, RESBLDG_ZIP)

    if not PARCEL_ZIP.exists():
        tmp_parcel = Path("/tmp/kc_parcel.zip")
        if tmp_parcel.exists():
            print(f"Using {tmp_parcel}")
            CACHE_DIR.mkdir(parents=True, exist_ok=True)
            import shutil
            shutil.copy(tmp_parcel, PARCEL_ZIP)
        else:
            download(PARCEL_URL, PARCEL_ZIP)


def stream_csv_from_zip(zip_path: Path, filename_hint: str = None):
    with zipfile.ZipFile(zip_path) as zf:
        names = zf.namelist()
        csv_name = names[0] if not filename_hint else next(n for n in names if filename_hint in n)
        with zf.open(csv_name) as f:
            import io
            reader = csv.DictReader(io.TextIOWrapper(f, encoding="utf-8", errors="replace"))
            for i, row in enumerate(reader):
                if i % 200_000 == 0 and i > 0:
                    print(f"  {i:,} rows...", flush=True)
                yield row


def main():
    ensure_cached()

    # Step 1: Read ResBldg — collect zip, yrBuilt, hasGarage per parcel (Major key)
    print("\nReading Residential Building table...")
    resbldg: dict[str, dict] = {}
    for row in stream_csv_from_zip(RESBLDG_ZIP):
        major = row.get("Major", "").strip()
        if not major:
            continue
        yr = int(row.get("YrBuilt", "0") or 0)
        gar_att  = int(row.get("SqFtGarageAttached",  "0") or 0)
        gar_bsmt = int(row.get("SqFtGarageBasement",  "0") or 0)
        zip_code = (row.get("ZipCode", "") or "").strip().strip('"')
        if yr > 1800:
            resbldg[major] = {
                "zip":      zip_code,
                "yr":       yr,
                "garage":   (gar_att + gar_bsmt) > 0,
            }
    print(f"  Loaded {len(resbldg):,} residential building records")

    # Step 2: Read Parcel — collect sqFtLot, presentUse per Major
    print("\nReading Parcel table...")
    parcels: dict[str, dict] = {}
    for row in stream_csv_from_zip(PARCEL_ZIP):
        major = row.get("Major", "").strip()
        if not major:
            continue
        lot = int(row.get("SqFtLot", "0") or 0)
        use = (row.get("PresentUse", "") or "").strip()
        parcels[major] = {"lot": lot, "use": use}
    print(f"  Loaded {len(parcels):,} parcel records")

    # Step 3: Aggregate per neighborhood
    agg: dict[str, dict] = {
        n: {"yrs": [], "lots": [], "garage": 0, "sfh": 0, "th": 0, "condo": 0, "total": 0}
        for n in ALL_NEIGHBORHOODS
    }

    skipped_zip = 0
    skipped_use = 0

    for major, bldg in resbldg.items():
        zip_code = bldg["zip"]
        neighborhoods = ZIP_MAP.get(zip_code)
        if not neighborhoods:
            skipped_zip += 1
            continue

        parcel = parcels.get(major, {"lot": 0, "use": ""})
        use = parcel["use"]

        is_sfh     = use in USE_SINGLE_FAMILY
        is_th      = use in USE_TOWNHOUSE
        is_condo   = use in USE_CONDO

        if not (is_sfh or is_th or is_condo):
            skipped_use += 1
            continue

        for name in neighborhoods:
            d = agg[name]
            d["total"] += 1
            d["yrs"].append(bldg["yr"])
            lot = parcel["lot"]
            if 100 < lot < 100_000:  # filter out zero/implausible values
                d["lots"].append(lot)
            if bldg["garage"]:
                d["garage"] += 1
            if is_sfh:   d["sfh"]   += 1
            if is_th:    d["th"]    += 1
            if is_condo: d["condo"] += 1

    print(f"\n  Skipped {skipped_zip:,} records (ZIP not in Seattle neighborhoods)")
    print(f"  Skipped {skipped_use:,} records (non-residential use code)")

    # Step 4: Compute stats
    result = {}
    for name in ALL_NEIGHBORHOODS:
        d = agg[name]
        n = d["total"]
        if n == 0:
            print(f"  WARN: no data for {name}")
            continue

        med_yr  = int(stat_median(d["yrs"])) if d["yrs"] else None
        med_lot = int(stat_median(d["lots"])) if d["lots"] else None

        def pct(k: int) -> int:
            return round(k / n * 100)

        result[name] = {
            "medianYearBuilt":  med_yr,
            "medianLotSqft":    med_lot,
            "pctWithGarage":    pct(d["garage"]),
            "housingMix": {
                "pctSingleFamily": pct(d["sfh"]),
                "pctTownhome":     pct(d["th"]),
                "pctCondo":        pct(d["condo"]),
            },
            "_sampleSize": n,
            "_note": "King County Assessor EXTR_ResBldg.csv + EXTR_Parcel.csv (May 2026)",
        }

    # Step 5: Print + write
    print("\n── Results ──────────────────────────────────────────────────────────")
    print(f"{'Neighborhood':<28} {'Built':>6} {'Lot':>7} {'Gar%':>5} {'SFH%':>5} {'TH%':>4} {'Condo%':>7} {'n':>6}")
    print("─" * 80)
    for name, d in sorted(result.items()):
        mix = d["housingMix"]
        print(
            f"{name:<28} {str(d['medianYearBuilt']):>6} {str(d['medianLotSqft']):>7} "
            f"{d['pctWithGarage']:>4}% {mix['pctSingleFamily']:>4}% {mix['pctTownhome']:>3}% "
            f"{mix['pctCondo']:>6}%  {d['_sampleSize']:>6,}"
        )

    with open(OUTPUT, "w") as f:
        json.dump(result, f, indent=2)
    print(f"\nWrote {OUTPUT}")


if __name__ == "__main__":
    main()

File: scripts/process-parcel-data.mjs
#!/usr/bin/env node
/**
 * process-parcel-data.mjs
 *
 * Computes housing characteristics for Seattle neighborhoods from King County
 * Assessor parcel data. Downloads data if not already cached locally.
 *
 * Sources:
 *   - King County Assessor: EXTR_ResBldg.csv  (year built, garage sqft, zip code)
 *   - King County Assessor: EXTR_Parcel.csv   (lot size, present use code)
 *
 * Usage:
 *   node scripts/process-parcel-data.mjs
 *
 * Outputs:
 *   scripts/housing-characteristics.json
 */

import { createReadStream, createWriteStream, existsSync, statSync, mkdirSync } from 'fs';
import { readFile, writeFile } from 'fs/promises';
import { createInterface } from 'readline';
import { createUnzip } from 'zlib';
import https from 'https';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CACHE_DIR = path.join(__dirname, '.cache');
const RESBLDG_URL = 'https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip';
const PARCEL_URL  = 'https://aqua.kingcounty.gov/extranet/assessor/Parcel.zip';
const RESBLDG_ZIP = path.join(CACHE_DIR, 'EXTR_ResBldg.zip');
const PARCEL_ZIP  = path.join(CACHE_DIR, 'EXTR_Parcel.zip');
const OUTPUT_FILE = path.join(__dirname, 'housing-characteristics.json');

// King County Parcel PresentUse codes (from LookUp table, LUType 102)
const USE_SINGLE_FAMILY = new Set(['2', '6', '9']);
const USE_TOWNHOUSE     = new Set(['29']);
const USE_CONDO         = new Set(['20', '25']);

// ZIP code → neighborhood mapping
// Where a ZIP spans multiple neighborhoods, it's noted. Parcels are assigned
// to the neighborhood listed (the dominant one for that ZIP in Seattle proper).
// A few neighborhoods share ZIPs; those are marked with a note.
const ZIP_TO_NEIGHBORHOOD = {
  '98109': 'Queen Anne',
  '98119': 'Queen Anne',     // Upper Queen Anne / west slope
  '98107': 'Ballard',
  '98117': 'Crown Hill',     // Crown Hill / NW Seattle (Ballard overlaps but 98107 covers Ballard core)
  // Sunset Hill uses same 98177 zip as Magnolia north but is geographically NW
  // We assign 98177 to Sunset Hill since Magnolia is covered by 98199
  '98177': 'Sunset Hill',
  '98103': 'Wallingford',    // Also covers Fremont; we split by counting both
  '98105': 'Wallingford',    // University District edge bleeds in; majority is Wallingford/U-District
  '98115': 'Roosevelt',      // Includes Ravenna/Roosevelt/Maple Leaf north
  '98125': 'Maple Leaf / Ravenna',
  '98133': 'Phinney / Greenwood',
  '98199': 'Magnolia',
  '98112': 'Madison Valley', // Madison Valley / Capitol Hill east
  '98102': 'Capitol Hill',
  '98116': 'West Seattle (Admiral)',
  '98118': 'Columbia City',
  '98144': 'Mount Baker',
  '98108': 'Beacon Hill',
  '98106': 'Beacon Hill',    // north Beacon Hill / Georgetown
};

// Fremont and Green Lake share 98103 with Wallingford. Give them their own entries
// using 98103 as a shared ZIP — we'll split it by noting the overlap.
// For Fremont we also use zip 98103 (Fremont Ave area).
// In practice 98103 covers Fremont, Wallingford, and the south edge of Green Lake.
// We assign 98103 to both Fremont and Wallingford (each gets the full 98103 stats).
// Green Lake is better covered by 98115 north (overlaps) and 98103 south.
// This is a known limitation; we note it in the output.
const MULTI_NEIGHBORHOOD_ZIPS = {
  '98103': ['Wallingford', 'Fremont', 'Green Lake'],
};

// For neighborhoods that need secondary ZIPs
const EXTRA_ZIP_TO_NEIGHBORHOOD = {
  '98103': ['Wallingford', 'Fremont', 'Green Lake'],
};

// Median function
function median(arr) {
  if (arr.length === 0) return null;
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? Math.round((sorted[mid - 1] + sorted[mid]) / 2)
    : sorted[mid];
}

// Cache freshness: 30 days
const CACHE_MAX_AGE_DAYS = 30;
function isCacheStale(filePath) {
  if (!existsSync(filePath)) return true;
  const age = (Date.now() - statSync(filePath).mtimeMs) / (1000 * 60 * 60 * 24);
  return age > CACHE_MAX_AGE_DAYS;
}

function download(url, dest) {
  return new Promise((resolve, reject) => {
    const file = createWriteStream(dest);
    const request = (u) => {
      https.get(u, (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
          file.close();
          return request(res.headers.location);
        }
        if (res.statusCode !== 200) {
          reject(new Error(`HTTP ${res.statusCode} for ${u}`));
          return;
        }
        const total = parseInt(res.headers['content-length'] || '0', 10);
        let downloaded = 0;
        res.on('data', (chunk) => {
          downloaded += chunk.length;
          if (total > 0) {
            const pct = Math.round((downloaded / total) * 100);
            process.stdout.write(`\r  ${pct}% (${(downloaded / 1e6).toFixed(1)} MB)`);
          }
        });
        res.pipe(file);
        file.on('finish', () => { file.close(); process.stdout.write('\n'); resolve(); });
      }).on('error', reject);
    };
    request(url);
  });
}

// Streaming CSV line reader that handles quoted fields
function* parseCsvLine(line) {
  const fields = [];
  let field = '';
  let inQuote = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      inQuote = !inQuote;
    } else if (ch === ',' && !inQuote) {
      fields.push(field.trim());
      field = '';
    } else {
      field += ch;
    }
  }
  fields.push(field.trim());
  return fields;
}

async function streamCsvFromZip(zipPath, callback) {
  return new Promise((resolve, reject) => {
    const { execSync } = await import('child_process').then(m => m);
    // Use unzip -p to stream from zip
    const { spawn } = require('child_process');
    const proc = spawn('unzip', ['-p', zipPath]);
    const unzip = proc.stdout;
    const rl = createInterface({ input: unzip });
    let header = null;
    let lineCount = 0;
    rl.on('line', (line) => {
      lineCount++;
      if (lineCount % 100000 === 0) process.stdout.write(`\r  ${(lineCount / 1e6).toFixed(1)}M rows...`);
      const fields = [...parseCsvLine(line)];
      if (!header) { header = fields; return; }
      const row = {};
      header.forEach((h, i) => { row[h] = fields[i] || ''; });
      callback(row);
    });
    rl.on('close', () => { if (lineCount > 1) process.stdout.write('\n'); resolve(); });
    proc.on('error', reject);
  });
}

async function main() {
  mkdirSync(CACHE_DIR, { recursive: true });

  // Download if stale
  if (isCacheStale(RESBLDG_ZIP)) {
    console.log('Downloading King County Residential Building data...');
    await download(RESBLDG_URL, RESBLDG_ZIP);
  } else {
    console.log('Using cached Residential Building data');
  }
  if (isCacheStale(PARCEL_ZIP)) {
    console.log('Downloading King County Parcel data...');
    await download(PARCEL_URL, PARCEL_ZIP);
  } else {
    console.log('Using cached Parcel data');
  }

  // Build structures
  // resbldg: Major → { zipCode, yrBuilt, garageAttached, garageBasement }
  console.log('\nReading Residential Building table...');
  const resbldg = new Map();
  await streamCsvFromZip(RESBLDG_ZIP, (row) => {
    const major = row['Major'];
    const zip = (row['ZipCode'] || '').trim();
    const yrBuilt = parseInt(row['YrBuilt'] || '0', 10);
    const garAtt = parseInt(row['SqFtGarageAttached'] || '0', 10);
    const garBsmt = parseInt(row['SqFtGarageBasement'] || '0', 10);
    if (major && yrBuilt > 1800) {
      resbldg.set(major, { zip, yrBuilt, hasGarage: (garAtt + garBsmt) > 0 });
    }
  });

  // parcel: Major → { sqFtLot, presentUse }
  console.log('Reading Parcel table...');
  const parcels = new Map();
  await streamCsvFromZip(PARCEL_ZIP, (row) => {
    const major = row['Major'];
    const sqFtLot = parseInt(row['SqFtLot'] || '0', 10);
    const use = (row['PresentUse'] || '').trim();
    if (major) {
      parcels.set(major, { sqFtLot, use });
    }
  });

  // Aggregate per neighborhood
  const data = {};
  const allNeighborhoods = new Set(Object.values(ZIP_TO_NEIGHBORHOOD));
  Object.values(EXTRA_ZIP_TO_NEIGHBORHOOD).flat().forEach(n => allNeighborhoods.add(n));
  allNeighborhoods.forEach(n => {
    data[n] = { yrBuilts: [], lotSizes: [], garageCount: 0, total: 0, sfh: 0, townhome: 0, condo: 0 };
  });

  for (const [major, bldg] of resbldg) {
    const zip = bldg.zip;
    let neighborhoods = [];
    if (EXTRA_ZIP_TO_NEIGHBORHOOD[zip]) {
      neighborhoods = EXTRA_ZIP_TO_NEIGHBORHOOD[zip];
    } else if (ZIP_TO_NEIGHBORHOOD[zip]) {
      neighborhoods = [ZIP_TO_NEIGHBORHOOD[zip]];
    }
    if (neighborhoods.length === 0) continue;

    const parcel = parcels.get(major) || { sqFtLot: 0, use: '' };
    const use = parcel.use;

    // Only count residential use codes — skip commercial/vacant
    const isResidential = USE_SINGLE_FAMILY.has(use) || USE_TOWNHOUSE.has(use) || USE_CONDO.has(use)
      || use === '2' || use === '29' || use === '20';
    if (!isResidential) continue;

    for (const name of neighborhoods) {
      if (!data[name]) continue;
      const d = data[name];
      d.total++;
      if (bldg.yrBuilt > 1800) d.yrBuilts.push(bldg.yrBuilt);
      if (parcel.sqFtLot > 0 && parcel.sqFtLot < 100000) d.lotSizes.push(parcel.sqFtLot);
      if (bldg.hasGarage) d.garageCount++;
      if (USE_SINGLE_FAMILY.has(use)) d.sfh++;
      else if (USE_TOWNHOUSE.has(use)) d.townhome++;
      else if (USE_CONDO.has(use)) d.condo++;
    }
  }

  // Compute final stats
  const result = {};
  for (const [name, d] of Object.entries(data)) {
    if (d.total === 0) { console.warn(`  WARN: no data for ${name}`); continue; }
    const pct = (n) => Math.round((n / d.total) * 100);
    result[name] = {
      medianYearBuilt:  median(d.yrBuilts) || null,
      medianLotSqft:    median(d.lotSizes) || null,
      pctWithGarage:    d.total > 0 ? pct(d.garageCount) : null,
      housingMix: {
        pctSingleFamily: pct(d.sfh),
        pctTownhome:     pct(d.townhome),
        pctCondo:        pct(d.condo),
      },
      _sampleSize: d.total,
      _source: 'King County Assessor EXTR_ResBldg.csv + EXTR_Parcel.csv (May 2026)',
    };
  }

  await writeFile(OUTPUT_FILE, JSON.stringify(result, null, 2));
  console.log(`\nWrote ${OUTPUT_FILE}`);
  console.log('\nResults:');
  for (const [name, d] of Object.entries(result)) {
    console.log(`  ${name.padEnd(28)} built:${d.medianYearBuilt}  lot:${d.medianLotSqft}sqft  garage:${d.pctWithGarage}%  sfh:${d.housingMix.pctSingleFamily}%  th:${d.housingMix.pctTownhome}%  condo:${d.housingMix.pctCondo}%  (n=${d._sampleSize})`);
  }
}

main().catch(e => { console.error(e); process.exit(1); });
